### PR TITLE
Add caching for Veldrid staging buffers in favour of `CommandList.UpdateBuffer`

### PR DESCRIPTION
**Pull request by frenzibyte**
- [x] Depends on #5911 (for caching logic)

Staging buffer cache has been removed from Metal in https://github.com/ppy/veldrid/pull/31, and needs to be brought back here to maintain performance. Needs testing on all platforms, especially D3D11 since it's the only backend that doesn't allocate a staging buffer in `CommandList.UpdateBuffer` (if there's any sign of performance degradation, I can easily revert back just for D3D11).

### CODE REVIEW

**Comment by smoogipoo** [comment]
I guess this may be able to replace the other staging buffer implementation (`Renderer.CreateStagingBuffer`)?

**Comment by frenzibyte** [comment]
Sure, could also change the pool to cache `IStagingBuffer` resources instead.

**Comment by frenzibyte** [comment]
Actually, there are usages of `CreateStagingBuffer` which permanently keep the staging buffer like VBOs, so we can't replace that. 

Also, changing the pool to return `IStagingBuffer`s instead gets a bit weird because that component is made to be initialised ahead of time, rather than being requested after a GPU buffer was set up already. Here's a diff showcasing the changes required:
```diff
diff --git a/osu.Framework/Graphics/Veldrid/Buffers/Staging/IStagingBuffer.cs b/osu.Framework/Graphics/Veldrid/Buffers/Staging/IStagingBuffer.cs
index a294254b9..34a107f89 100644
--- a/osu.Framework/Graphics/Veldrid/Buffers/Staging/IStagingBuffer.cs
+++ b/osu.Framework/Graphics/Veldrid/Buffers/Staging/IStagingBuffer.cs
@@ -6,8 +6,7 @@
 
 namespace osu.Framework.Graphics.Veldrid.Buffers.Staging
 {
-    internal interface IStagingBuffer<T> : IDisposable
-        where T : unmanaged
+    internal interface IStagingBuffer : IDisposable
     {
         /// <summary>
         /// The total size of this buffer in bytes.
@@ -24,18 +23,22 @@ internal interface IStagingBuffer<T> : IDisposable
         /// </summary>
         BufferUsage CopyTargetUsageFlags { get; }
 
-        /// <summary>
-        /// The data contained in this buffer.
-        /// </summary>
-        Span<T> Data { get; }
-
         /// <summary>
         /// Copies data from this buffer into a <see cref="DeviceBuffer"/>.
         /// </summary>
         /// <param name="buffer">The target buffer.</param>
         /// <param name="srcOffset">The offset into this buffer at which the copy should start.</param>
         /// <param name="dstOffset">The offset into <paramref name="buffer"/> at which the copy should start.</param>
-        /// <param name="size">The number of elements to be copied.</param>
-        void CopyTo(DeviceBuffer buffer, uint srcOffset, uint dstOffset, uint size);
+        /// <param name="count">The number of elements to be copied.</param>
+        void CopyTo(DeviceBuffer buffer, uint srcOffset, uint dstOffset, uint count);
+    }
+
+    internal interface IStagingBuffer<T> : IStagingBuffer
+        where T : unmanaged
+    {
+        /// <summary>
+        /// The data contained in this buffer.
+        /// </summary>
+        Span<T> Data { get; }
     }
 }
diff --git a/osu.Framework/Graphics/Veldrid/Buffers/VeldridIndexData.cs b/osu.Framework/Graphics/Veldrid/Buffers/VeldridIndexData.cs
index dae7ae1e1..f32043b75 100644
--- a/osu.Framework/Graphics/Veldrid/Buffers/VeldridIndexData.cs
+++ b/osu.Framework/Graphics/Veldrid/Buffers/VeldridIndexData.cs
@@ -5,35 +5,24 @@
 
 namespace osu.Framework.Graphics.Veldrid.Buffers
 {
-    // todo: this feels too janky
     internal class VeldridIndexData
     {
-        private readonly VeldridRenderer renderer;
-
-        private int capacity;
-
-        public int Capacity
-        {
-            get => capacity;
-            set
-            {
-                if (value == capacity)
-                    return;
-
-                capacity = value;
+        public int Capacity { get; private set; }
+        public DeviceBuffer Buffer { get; private set; }
 
-                buffer?.Dispose();
-                buffer = null;
-            }
-        }
+        private readonly VeldridRenderer renderer;
 
         public VeldridIndexData(VeldridRenderer renderer)
         {
             this.renderer = renderer;
         }
 
-        private DeviceBuffer? buffer;
+        public void Recreate(int capacity, BufferUsage usageFlags)
+        {
+            Buffer?.Dispose();
+            Buffer = renderer.Factory.CreateBuffer(new BufferDescription((uint)(capacity * sizeof(ushort)), BufferUsage.IndexBuffer | usageFlags));
 
-        public DeviceBuffer Buffer => buffer ??= renderer.Factory.CreateBuffer(new BufferDescription((uint)(Capacity * sizeof(ushort)), BufferUsage.IndexBuffer));
+            Capacity = capacity;
+        }
     }
 }
diff --git a/osu.Framework/Graphics/Veldrid/Buffers/VeldridLinearBuffer.cs b/osu.Framework/Graphics/Veldrid/Buffers/VeldridLinearBuffer.cs
index 733756020..98d843343 100644
--- a/osu.Framework/Graphics/Veldrid/Buffers/VeldridLinearBuffer.cs
+++ b/osu.Framework/Graphics/Veldrid/Buffers/VeldridLinearBuffer.cs
@@ -30,16 +30,13 @@ protected override void Initialise()
 
             if (amountVertices > renderer.SharedLinearIndex.Capacity)
             {
-                renderer.SharedLinearIndex.Capacity = amountVertices;
-
-                ushort[] indices = new ushort[amountVertices];
+                var staging = renderer.GetFreeStagingBuffer<ushort>(amountVertices);
+                renderer.SharedLinearIndex.Recreate(amountVertices, staging.CopyTargetUsageFlags);
 
                 for (ushort i = 0; i < amountVertices; i++)
-                    indices[i] = i;
+                    staging.Data[i] = i;
 
-                var staging = renderer.GetFreeStagingBuffer(renderer.SharedLinearIndex.Buffer.SizeInBytes);
-                renderer.Device.UpdateBuffer(staging, 0, indices);
-                renderer.BufferUpdateCommands.CopyBuffer(staging, 0, renderer.SharedLinearIndex.Buffer, 0, staging.SizeInBytes);
+                staging.CopyTo(renderer.SharedLinearIndex.Buffer, 0, 0, (uint)amountVertices);
             }
         }
 
diff --git a/osu.Framework/Graphics/Veldrid/Buffers/VeldridQuadBuffer.cs b/osu.Framework/Graphics/Veldrid/Buffers/VeldridQuadBuffer.cs
index 663fbd214..f2d7158aa 100644
--- a/osu.Framework/Graphics/Veldrid/Buffers/VeldridQuadBuffer.cs
+++ b/osu.Framework/Graphics/Veldrid/Buffers/VeldridQuadBuffer.cs
@@ -37,23 +37,20 @@ protected override void Initialise()
 
             if (amountIndices > renderer.SharedQuadIndex.Capacity)
             {
-                renderer.SharedQuadIndex.Capacity = amountIndices;
-
-                ushort[] indices = new ushort[amountIndices];
+                var staging = renderer.GetFreeStagingBuffer<ushort>(amountIndices);
+                renderer.SharedQuadIndex.Recreate(amountIndices, staging.CopyTargetUsageFlags);
 
                 for (ushort i = 0, j = 0; j < amountIndices; i += IRenderer.VERTICES_PER_QUAD, j += indices_per_quad)
                 {
-                    indices[j] = i;
-                    indices[j + 1] = (ushort)(i + 1);
-                    indices[j + 2] = (ushort)(i + 3);
-                    indices[j + 3] = (ushort)(i + 2);
-                    indices[j + 4] = (ushort)(i + 3);
-                    indices[j + 5] = (ushort)(i + 1);
+                    staging.Data[j] = i;
+                    staging.Data[j + 1] = (ushort)(i + 1);
+                    staging.Data[j + 2] = (ushort)(i + 3);
+                    staging.Data[j + 3] = (ushort)(i + 2);
+                    staging.Data[j + 4] = (ushort)(i + 3);
+                    staging.Data[j + 5] = (ushort)(i + 1);
                 }
 
-                var staging = renderer.GetFreeStagingBuffer(renderer.SharedQuadIndex.Buffer.SizeInBytes);
-                renderer.Device.UpdateBuffer(staging, 0, indices);
-                renderer.BufferUpdateCommands.CopyBuffer(staging, 0, renderer.SharedQuadIndex.Buffer, 0, staging.SizeInBytes);
+                staging.CopyTo(renderer.SharedQuadIndex.Buffer, 0, 0, (uint)amountIndices);
             }
         }
 
diff --git a/osu.Framework/Graphics/Veldrid/Buffers/VeldridUniformBufferStorage.cs b/osu.Framework/Graphics/Veldrid/Buffers/VeldridUniformBufferStorage.cs
index 1e6b5035a..bc3ce17c7 100644
--- a/osu.Framework/Graphics/Veldrid/Buffers/VeldridUniformBufferStorage.cs
+++ b/osu.Framework/Graphics/Veldrid/Buffers/VeldridUniformBufferStorage.cs
@@ -12,8 +12,9 @@ internal class VeldridUniformBufferStorage<TData>
         where TData : unmanaged, IEquatable<TData>
     {
         private readonly VeldridRenderer renderer;
-        private readonly DeviceBuffer buffer;
-        private readonly NativeMemoryTracker.NativeMemoryLease memoryLease;
+
+        private DeviceBuffer? buffer;
+        private NativeMemoryTracker.NativeMemoryLease? memoryLease;
 
         private ResourceSet? set;
         private TData data;
@@ -21,9 +22,6 @@ internal class VeldridUniformBufferStorage<TData>
         public VeldridUniformBufferStorage(VeldridRenderer renderer)
         {
             this.renderer = renderer;
-
-            buffer = renderer.Factory.CreateBuffer(new BufferDescription((uint)Marshal.SizeOf(default(TData)), BufferUsage.UniformBuffer));
-            memoryLease = NativeMemoryTracker.AddMemory(this, buffer.SizeInBytes);
         }
 
         public TData Data
@@ -33,9 +31,16 @@ public TData Data
             {
                 data = value;
 
-                var staging = renderer.GetFreeStagingBuffer(buffer.SizeInBytes);
-                renderer.Device.UpdateBuffer(staging, 0, ref data);
-                renderer.BufferUpdateCommands.CopyBuffer(staging, 0, buffer, 0, buffer.SizeInBytes);
+                var staging = renderer.GetFreeStagingBuffer<TData>(1);
+
+                if (buffer == null)
+                {
+                    buffer = renderer.Factory.CreateBuffer(new BufferDescription((uint)Marshal.SizeOf(default(TData)), BufferUsage.UniformBuffer | staging.CopyTargetUsageFlags));
+                    memoryLease = NativeMemoryTracker.AddMemory(this, buffer.SizeInBytes);
+                }
+
+                staging.Data[0] = data;
+                staging.CopyTo(buffer, 0, 0, 1);
             }
         }
 
@@ -43,8 +48,8 @@ public TData Data
 
         public void Dispose()
         {
-            buffer.Dispose();
-            memoryLease.Dispose();
+            buffer?.Dispose();
+            memoryLease?.Dispose();
             set?.Dispose();
         }
     }
diff --git a/osu.Framework/Graphics/Veldrid/VeldridRenderer.cs b/osu.Framework/Graphics/Veldrid/VeldridRenderer.cs
index 00089ba1b..f3e6908f3 100644
--- a/osu.Framework/Graphics/Veldrid/VeldridRenderer.cs
+++ b/osu.Framework/Graphics/Veldrid/VeldridRenderer.cs
@@ -676,11 +676,15 @@ internal IStagingBuffer<T> CreateStagingBuffer<T>(uint count)
             }
         }
 
-        protected override void SetUniformImplementation<T>(IUniformWithValue<T> uniform)
+        public IStagingBuffer<T> GetFreeStagingBuffer<T>(int count)
+            where T : unmanaged
         {
+            return stagingBufferPool.Get<T>(count);
         }
 
-        public DeviceBuffer GetFreeStagingBuffer(uint sizeInBytes) => stagingBufferPool.Get(sizeInBytes);
+        protected override void SetUniformImplementation<T>(IUniformWithValue<T> uniform)
+        {
+        }
 
         public void RegisterUniformBufferForReset(IVeldridUniformBuffer buffer)
         {
diff --git a/osu.Framework/Graphics/Veldrid/VeldridStagingBufferPool.cs b/osu.Framework/Graphics/Veldrid/VeldridStagingBufferPool.cs
index 8c772a91f..18cc9d241 100644
--- a/osu.Framework/Graphics/Veldrid/VeldridStagingBufferPool.cs
+++ b/osu.Framework/Graphics/Veldrid/VeldridStagingBufferPool.cs
@@ -1,25 +1,26 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
-using Veldrid;
+using osu.Framework.Graphics.Veldrid.Buffers.Staging;
 
 namespace osu.Framework.Graphics.Veldrid
 {
-    internal class VeldridStagingBufferPool : VeldridStagingResourcePool<DeviceBuffer>
+    internal class VeldridStagingBufferPool : VeldridStagingResourcePool<IStagingBuffer>
     {
         public VeldridStagingBufferPool(VeldridRenderer renderer)
             : base(renderer, "staging buffers")
         {
         }
 
-        public DeviceBuffer Get(uint sizeInBytes)
+        public IStagingBuffer<T> Get<T>(int count)
+            where T : unmanaged
         {
-            if (TryGet(b => b.SizeInBytes >= sizeInBytes, out DeviceBuffer? buffer))
-                return buffer;
+            if (TryGet(b => b is IStagingBuffer<T> && b.Count >= count, out IStagingBuffer? buffer))
+                return (IStagingBuffer<T>)buffer;
 
-            buffer = Renderer.Factory.CreateBuffer(new BufferDescription(sizeInBytes, BufferUsage.Staging));
+            buffer = Renderer.CreateStagingBuffer<T>((uint)count);
             AddNewResource(buffer);
-            return buffer;
+            return (IStagingBuffer<T>)buffer;
         }
     }
 }
diff --git a/osu.Framework/Graphics/Veldrid/VeldridStagingResourcePool.cs b/osu.Framework/Graphics/Veldrid/VeldridStagingResourcePool.cs
index f4ac0964c..0b8c0ecb8 100644
--- a/osu.Framework/Graphics/Veldrid/VeldridStagingResourcePool.cs
+++ b/osu.Framework/Graphics/Veldrid/VeldridStagingResourcePool.cs
@@ -5,12 +5,11 @@
 using System.Collections.Generic;
 using System.Diagnostics.CodeAnalysis;
 using osu.Framework.Statistics;
-using Veldrid;
 
 namespace osu.Framework.Graphics.Veldrid
 {
     internal abstract class VeldridStagingResourcePool<T>
-        where T : class, DeviceResource, IDisposable
+        where T : class, IDisposable
     {
         protected readonly VeldridRenderer Renderer;
 

```

This really feels like we should have an `IBuffer` concept instead, providing flags suggesting whether it's dynamic, or should be backed by a staging buffer, or should rent a staging buffer on upload, etc. But I don't want to touch on that furthermore in this PR as this PR intends to not change behaviour as much as possible, I can revert this change on D3D11 if that would feel any better (since it's the only backend that actually does native work in `CommandList.UpdateBuffer`, all other backends just rent a staging buffer and performs a copy).

**Comment by smoogipoo** [comment]
> `IBuffer`

Without having looked at the above diff, I'd agree with that direction.

**Comment by peppy** [comment]
@frenzibyte please check changes make sense to you.

**Review by peppy — DISMISSED** [review]

**Comment by frenzibyte** [comment]
All looking pretty A+ to me 👍🏻

**Review by peppy — APPROVED** [review]

**Comment by EVAST9919** [comment]
There's a big performance hit on D3D
Before pr:
before
After:
after

**Comment by peppy** [comment]
@EVAST9919 can you repro on a framework test scene? maybe one of the new performance ones? Or can you make one based on the edge case? I couldn't see any drop at my end in limited testing.

**Comment by peppy** [comment]
@EVAST9919 can you test https://github.com/ppy/osu-framework/pull/5916 ? Does it show the regression?
